feature/#22-categories-crud-operations-in-domain-and-data

iZyadNasser opened the pull request.

iZyadNasser committed:
feat: add category operations

alaafawzyy committed:
feat :add category dao operation and change abstract class to interface

alaafawzyy committed:
feat :add di for category dao and task dao

alaafawzyy committed:
feat :change return type for get category fun

alaafawzyy committed:
feat :add implementation for task service impl

iZyadNasser committed:
feat: add tasks services to di and rename data module

iZyadNasser committed:
fix: change tasks service function types to room entity instead of domain entity and add mapper from domain to room entity

aya-seif commented on app/src/main/java/com/moscow/tudee/data/local/dao/CategoryDao.kt:
```diff
@@ -1,8 +1,23 @@
-abstract class CategoryDao {
+interface CategoryDao {
+    @Query("SELECT * FROM categories_table")
+    suspend fun getCategories(): List<CategoryEntity>?
```
In case there is no categories it will return empty list not null

iZyadNasser replied:
ok I removed the nullability, thanks for the feedback!

FaresM0hamed commented on app/src/main/java/com/moscow/tudee/data/local/mapper/mapper.kt:
```diff
@@ -14,6 +14,15 @@ fun CategoryEntity.toCategory(): Category {
+        title = category.title,
+        iconUrl = category.iconUrl
+    )
+}
```
I think this fun need enhancement in name , because when i read it , i don't understand directly what is the mean of Entity here , so what do  you think ?

iZyadNasser replied:
I changed the name to getRoomEntityFromCategory, is it ok like that or do you have another suggestion?

devsadeq commented on app/src/main/java/com/moscow/tudee/data/local/mapper/mapper.kt:
```diff
@@ -14,6 +14,15 @@ fun CategoryEntity.toCategory(): Category {
+const val DUMMY_VALUE = 0L
+fun getEntityFromCategory(category: Category): CategoryEntity {
+    return CategoryEntity(
+        id = DUMMY_VALUE,
```
Why the id is always zero?

AhmadKharfan replied:
I think because he use autoGenerate id in the entity so Room will ignore the 0L value during insertion and assign a new auto generated ID

iZyadNasser replied:
the id is auto generated by Room, it is the responsibility of the data layer not the domain so I just needed a place holder and couldn't find any other "more elegant" way to put a dummy temporary value, if you have a better approach, I'd be happy to know it

poula22 replied:
@iZyadNasser make it default value in class better I think as for readability now I thought you gave the 0 value id for all as constant, but in entity when I see default value with annotation Auto Generated it makes sense

iZyadNasser committed:
fix: github reviews
- remove nullability from the get categories functions return type
- change the domain category to room category mapper function name to be more specific

AhmadKharfan commented on app/src/main/java/com/moscow/tudee/data/local/dao/CategoryDao.kt:
```diff
@@ -1,8 +1,23 @@
-abstract class CategoryDao {
+interface CategoryDao {
+    @Query("SELECT * FROM categories_table")
+    suspend fun getCategories(): List<CategoryEntity>
```
I think it's better to return the list of categories as Flow 
This way any changes to the underlying data will be emitted automatically

FaresM0hamed reviewed (approved):
Good job

poula22 commented on app/src/main/java/com/moscow/tudee/data/service/TasksServicesImpl.kt:
```diff
@@ -1,7 +1,26 @@
+    }
+
+    override suspend fun addCategory(category: Category) {
+        categoryDao.addCategory(getRoomEntityFromCategory(category))
```
this can be replaced with mapper extension function getRoomEntityFromCategory(category)

poula22 commented on app/src/main/java/com/moscow/tudee/domain/service/TasksServices.kt:
```diff
@@ -1,4 +1,14 @@
 interface TasksServices {
+    suspend fun getCategories(): List<Category>
+
+    suspend fun addCategory(category: Category)
+
+    suspend fun updateCategory(category: Category)
+
+    suspend fun deleteCategory(categoryId: Long)
```
return type shouldn't be Unit retun Boolean if available to deal with negative action if needed

poula22 reviewed (commented):
It's better to make these changes if you have time, but not urgent
